### PR TITLE
NEXUS-8043: Multiple fixes for rebuild npm md task

### PR DESCRIPTION
**Pull request by cstamas**
Changes:
* fix for invalid package JSON, fallback to ISO8859-1
* fix for NPE, log statement wrongly located outside of null-check
* fix for various Version related exceptions, dropped Version, using String + alphanum comparator

Issue
https://issues.sonatype.org/browse/NEXUS-8043

CI
http://bamboo.s/browse/NX-OSSF597

### CODE REVIEW

File: plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/hosted/RecreateMetadataWalkerProcessor.java [hunk 1]
```diff
@@ -191,9 +193,21 @@ public void onCollectionExit(final WalkerContext context, final StorageCollectio
...
-          .readValue(bos.toByteArray(), new TypeReference<HashMap<String, Object>>() { });
-      return packageJson;
+
+      try {
+        return objectMapper.readValue(bos.toByteArray(), new TypeReference<HashMap<String, Object>>() { });
+      }
+      catch (JsonParseException e) {
+        // fallback
+        if (e.getMessage().contains("Invalid UTF-8 middle byte")) {
+          log.debug("JSON UTF-8 problem for package {} : {}, attempting with charset ISO-8859-2...",
```
Comment (kellyrob99) [inline plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/hosted/RecreateMetadataWalkerProcessor.java]: wrong charset mentioned, and not sure why we use ...?

File: plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/hosted/RecreateMetadataWalkerProcessor.java [hunk 2]
```diff
@@ -191,9 +193,21 @@ public void onCollectionExit(final WalkerContext context, final StorageCollectio
...
+              file.getRepositoryItemUid(), e.getMessage());
+          // try again, but assume ISO8859-1 encoding now, that is illegal for JSON
+          return objectMapper.readValue(new String(bos.toByteArray(), Charsets.ISO_8859_1),
```
Comment (kellyrob99) [inline plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/hosted/RecreateMetadataWalkerProcessor.java]: I wonder if this problem would be more easily resolved by doing this above and explicitly specifying UTF-8 instead?

File: plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/hosted/RecreateMetadataWalkerProcessor.java [hunk 3]
```diff
@@ -191,9 +193,21 @@ public void onCollectionExit(final WalkerContext context, final StorageCollectio
       // do this in memory, as package.json is expected to be fairly small (few kb at most)
       final ByteArrayOutputStream bos = new ByteArrayOutputStream(Ints.checkedCast(tarEntry.getSize()));
       StreamSupport.copy(tarArchiveInputStream, bos, StreamSupport.BUFFER_SIZE);
-      final Map<String, Object> packageJson = objectMapper
-          .readValue(bos.toByteArray(), new TypeReference<HashMap<String, Object>>() { });
-      return packageJson;
+
+      try {
+        return objectMapper.readValue(bos.toByteArray(), new TypeReference<HashMap<String, Object>>() { });
```
Comment (kellyrob99) [inline plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/hosted/RecreateMetadataWalkerProcessor.java]: Do we have an input that exhibits the 'bad' behaviour, and can we include that in a unit test?

File: plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/hosted/RecreateMetadataWalkerProcessor.java (hunk 1 above)
Comment (cstamas) [inline plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/hosted/RecreateMetadataWalkerProcessor.java]: reflex: ISO-8859-2 is Latin2, used in E-EU (and contains HU characters too for example)... :smile:  
Fix imminent.

File: plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/hosted/RecreateMetadataWalkerProcessor.java (hunk 2 above)
Comment (cstamas) [inline plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/hosted/RecreateMetadataWalkerProcessor.java]: No, the problem was that tarball contained package.json contains invalid UTF 8 character. It actually contains Latin1/ISO-8859-1 encoded character "é". It's exactly this case:
http://jackson-users.ning.com/forum/topics/utf-8-convertion-issues

File: plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/hosted/RecreateMetadataWalkerProcessor.java (hunk 3 above)
Comment (cstamas) [inline plugins/npm/nexus-npm-repository-plugin/src/main/java/com/bolyuba/nexus/plugin/npm/hosted/RecreateMetadataWalkerProcessor.java]: We do have, will add UT.

**Comment by kellyrob99** [comment]
+1, thanks @cstamas

**Pull request merged**